core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.JsonObject;
+import io.grpc.Attributes;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer.Helper;
+import io.grpc.LoadBalancer.PickResult;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.LoadBalancer.SubchannelPicker;
+import io.grpc.PickFirstBalancerFactory;
+import io.grpc.Status;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Locale;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class AutoConfiguredLoadBalancerFactory extends LoadBalancer.Factory {
+  private static final Logger logger =
+      Logger.getLogger(AutoConfiguredLoadBalancerFactory.class.getName());
+
+  @VisibleForTesting
+  static final String ROUND_ROUND_LOAD_BALANCER_FACTORY_NAME =
+      "io.grpc.util.RoundRobinLoadBalancerFactory";
+  @VisibleForTesting
+  static final String GRPCLB_LOAD_BALANCER_FACTORY_NAME =
+      "io.grpc.grpclb.GrpclbLoadBalancerFactory";
+
+  AutoConfiguredLoadBalancerFactory() {}
+
+  @Override
+  public LoadBalancer newLoadBalancer(Helper helper) {
+    return new AutoConfiguredLoadBalancer(helper);
+  }
+
+  private static final class EmptySubchannelPicker extends SubchannelPicker {
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      return PickResult.withNoResult();
+    }
+  }
+
+  static final class AutoConfiguredLoadBalancer extends LoadBalancer {
+    private final Helper helper;
+    private LoadBalancer delegate;
+    private LoadBalancer.Factory delegateFactory;
+
+    AutoConfiguredLoadBalancer(Helper helper) {
+      this.helper = helper;
+      delegateFactory = PickFirstBalancerFactory.getInstance();
+      delegate = delegateFactory.newLoadBalancer(helper);
+    }
+
+    //  Must be run inside ChannelExecutor.
+    @Override
+    public void handleResolvedAddressGroups(
+        List<EquivalentAddressGroup> servers, Attributes attributes) {
+      if (attributes.keys().contains(GrpcAttributes.NAME_RESOLVER_ATTR_SERVICE_CONFIG)) {
+        Factory newlbf = decideLoadBalancerFactory(
+            servers, attributes.get(GrpcAttributes.NAME_RESOLVER_ATTR_SERVICE_CONFIG));
+        if (newlbf != null && newlbf != delegateFactory) {
+          delegate.shutdown();
+          helper.updateBalancingState(ConnectivityState.CONNECTING, new EmptySubchannelPicker());
```

Comment:
I think updateBalancingState() should happen before delegate.shutdown().

delegate.shutdown() will shutdown all subchannels. RPCs started between shutdown() and updateBalancingState() will be assigned to shutdown'ed subchannels and will fail.